### PR TITLE
VUU86: Add tests for layout server

### PR DESCRIPTION
**Pull request by cfisher-scottlogic**
### Description
Part of #57.

The layout server implementation PR is [here](https://github.com/ScottLogic/finos-vuu/pull/57). This is the separate PR for all of the related tests.

It's definitely worth thinking about whether the edge cases are giving us the desired responses.

### Change List

- [x] `LayoutControllerTest` tests
- [x] `LayoutServiceTest` tests
- [x] `LayoutIntegrationTest` tests (tests endpoints with fully contextualized Spring server)
- [x] Add exception handling

### CODE REVIEW

File: layout-server/src/main/java/org/finos/vuu/layoutserver/controller/LayoutController.java [hunk 1]
```diff
@@ -98,6 +98,8 @@ public void updateLayout(@PathVariable UUID id, @RequestBody @Valid LayoutReques
     @ResponseStatus(HttpStatus.NO_CONTENT)
     @DeleteMapping("/{id}")
     public void deleteLayout(@PathVariable UUID id) {
+        // Generate a 404 if layout doesn't exist
+        layoutService.getLayout(id);
```
Comment (cfisher-scottlogic) [inline layout-server/src/main/java/org/finos/vuu/layoutserver/controller/LayoutController.java:102]: Do we want an exception to be thrown or do we want it to just silently do nothing?

**Comment by cfisher-scottlogic** [comment]
Leaving this here as a reminder to myself to address it in this PR:

> Let's make sure we test the transactional nature of this.
> https://github.com/ScottLogic/finos-vuu/pull/57#discussion_r1344149426

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        LayoutRequestDTO layoutRequest = createValidCreateRequest();
+        layoutRequest.setMetadata(null);
+
+        mockMvc.perform(post("/layouts")
+                .content(objectMapper.writeValueAsString(layoutRequest))
+                .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isBadRequest());
+
+        assertThat(layoutRepository.findAll()).isEmpty();
+    }
```
Comment (cfisher-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:139]: > Let's make sure we test the transactional nature of this.
> https://github.com/ScottLogic/finos-vuu/pull/57#discussion_r1344149426

@pling-scottlogic Do you think this test sufficiently covers this?

File: layout-server/src/main/java/org/finos/vuu/layoutserver/controller/LayoutController.java (hunk 1 above)
Comment (heswell) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/controller/LayoutController.java:101]: I'm inclined to say just logging it would be enough. DOn't see much value in informing the user - something has obviously gone wrong if this happens, so we should capture it, but its not going to affect the users experience.

**Review by heswell — COMMENTED** [review]
the layouts and metadata endpoints  look fine. There is one additional endpoint needed, which is the application level layout document.

This is basically going to be a list of the loaded layouts (which will be represented as the main tabs) , plus indication of which one is currently open.


Its what we currently hardcode as defaultLayout in the sample app.

FIrst call made on startup will be to get this application layout. The content of the main tabs will be references to layouts, so something liike this is what I imagined...

```
{
  id: "main-tabs",
  type: "Stack",
  props: {
    className: "vuuShell-mainTabs",
    TabstripProps: {
      allowAddTab: true,
      allowRenameTab: true,
      animateSelectionThumb: false,
      className: "vuuShellMainTabstrip",
      location: "main-tab",
    },
    preserve: true,
    active: 0,
  },
  children: [
    {
     title:"My Layout"
      type: "Layout",
      id: "layout-id-001",
    },
  ],
}
``` 

from this, the client would then request "/layouts/layout-id-001"

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        MetadataResponseDTO metadataResponse = getMetadataResponseDTO();
+        expectedLayoutResponse.setMetadata(metadataResponse);
+
+        expectedMetadataResponse = new ArrayList<>();
+        expectedMetadataResponse.add(metadataResponse);
+    }
+
+
+    @Test
+    void getLayout_validIdAndLayoutExists_returnsLayout() {
```
Comment (vferraro-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:90]: What do we mean by validId here? is it just an id for which we have a layout? in that case `getLayout_layoutExists_returnsLayout()` would descriptive enough

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 2 above)
Comment (pling-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: Yeah, that looks good to me. I think I would just do the same assertion on the metadata repository, for completeness.

**Review by vferraro-scottlogic — APPROVED** [review]

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Test
+    void getMetadata_noMetadataExists_returnsEmptyArray() {
+        when(metadataService.getMetadata()).thenReturn(List.of());
+        assertThat(layoutController.getMetadata()).isEmpty();
+    }
+
+    @Test
+    void createLayout_validLayout_returnsCreatedLayout() {
```
Comment (vferraro-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:124]: should we add a test case for createdLayout resulting in 400? i.e. when description is ""

Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:124]: That's a test case in `LayoutIntegrationTest`. Do you think it's beneficial to cover this scenario in a unit test as well?

Comment (vferraro-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:124]: makes sense to cover it in integration tests

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 2 above)
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [bf7278e](https://github.com/ScottLogic/finos-vuu/pull/68/commits/bf7278e3202694a62712feabaf406c379640e73a)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java (hunk 4 above)
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:124]: [45ff964](https://github.com/ScottLogic/finos-vuu/pull/68/commits/45ff96496048a1aee5b42cd1547c8e50e1f8a1ac)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+    @Test
+    void deleteLayout_validIdLayoutExists_returnsSuccessAndLayoutIsDeleted() throws Exception {
+        Layout layout = createDefaultLayoutInDatabase();
+
+        mockMvc.perform(get("/layouts/{id}", layout.getId())).andExpect(status().isOk());
+
+        mockMvc.perform(delete("/layouts/{id}", layout.getId())).andExpect(status().isNoContent());
+
+        mockMvc.perform(get("/layouts/{id}", layout.getId())).andExpect(status().isNotFound());
```
Comment (pling-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:327]: I would prefer to just call `mockMvc.perform` for behaviour under test, i.e. hitting the `delete` endpoint. We could do a simpler assertion using the repo instead of the gets before and after.

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]
There are quite a few alterations for an "add tests" PR - are these all things that have come up as a result of writing the tests?

Could also do with a few more "unhappy" paths for the service.

File: layout-server/src/main/java/org/finos/vuu/layoutserver/service/LayoutService.java [hunk 6]
```diff
@@ -13,15 +15,21 @@
...
+            .orElseThrow(() -> new NoSuchElementException("Layout with ID '" + id + "' not found"));
     }
 
     public Layout getLayoutByMetadataId(UUID id) {
         return layoutRepository.findLayoutByMetadataId(id);
     }
 
+    public List<Metadata> getMetadata() {
+        return metadataService.getMetadata();
+    }
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/service/LayoutService.java:31]: Well spotted. Looks like it's just a change I made before I made the change on the previous PR (to move this to the controller) which I've not cleaned up. I'll remove this here.

File: layout-server/src/main/java/org/finos/vuu/layoutserver/controller/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        return new ResponseEntity<>(ex.getMessage(),
+            org.springframework.http.HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(HttpMessageNotReadableException.class)
+    public ResponseEntity<String> handleHttpMessageNotReadableException(
+        HttpMessageNotReadableException ex) {
+        return new ResponseEntity<>(ex.getMessage(),
+            org.springframework.http.HttpStatus.BAD_REQUEST);
+    }
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/controller/GlobalExceptionHandler.java:43]: I originally had that, and then separated it out (in [this commit](https://github.com/ScottLogic/finos-vuu/pull/68/commits/7dd8233feeb0442ac1db22ded9df90f86f54b6e0)) in an effort to improve the error messages, where `MethodArgumentNotValidException` needed to handle the error message slightly differently.

It could look confusing if we have a generic `handleBadRequest` exception handler when it isn't the only exception handler returning a bad request response. But, I do have a generic `handleNotFound`, so arguably that's hypocritical.

I think if we have the 'generic' exception handlers at the top, and any individual exception handlers below, I can get on board with this suggestion.

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+import org.finos.vuu.layoutserver.service.MetadataService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.modelmapper.ModelMapper;
+
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:29]: AFAIK it's necessary for the `@Mock` &  `@InjectMocks` annotations.

See [here](https://stackoverflow.com/a/60308868).

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Test
+    void getMetadata_metadataExists_returnsMetadata() {
+        when(metadataService.getMetadata()).thenReturn(List.of(metadata));
+        when(modelMapper.map(metadata, MetadataResponseDTO.class)).thenReturn(
+            getMetadataResponseDTO());
+
+        assertThat(layoutController.getMetadata()).isEqualTo(expectedMetadataResponse);
+    }
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java:115]: Ah, I applied your suggestion and actually spotted that the test didn't look to be doing the right comparison anyway. Thanks!

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.transaction.annotation.Transactional;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@Transactional
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:38]: It's there to ensure data from one test does not affect the other by rolling back any database modifications. Do you know of a better way?

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    public static void setup() {
+        defaultDefinition = "Default layout definition";
+        defaultName = "Default layout name";
+        defaultGroup = "Default layout group";
+        defaultScreenshot = "Default layout screenshot";
+        defaultUser = "Default layout user";
+    }
+
+    @Test
+    void getLayout_validIDAndLayoutExists_returns200WithLayout() throws Exception {
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:67]: To be honest I autogenerate the boilerplate of my test files with intellij, and no access modifier is what it defaults to. I don't have a particular opinion on it, although I like [this answer](https://stackoverflow.com/a/55230350) because it involves less typing 😄

> > _Why is the default access modifier in JUnit 5 package-private?_
> 
> It's not the "default". There technically is no default. Rather, in JUnit Jupiter you have a choice: public, protected or package-private.
> 
> > _What is the benefit of changing it to package-private?_
> 
> The benefit is that you don't have type public anymore. If your IDE automatically generates test methods and test classes for you that are public, feel free to leave them public.
> 
> But... if you are typing in the methods on your own, then just leave off public unless you are designing your test classes for subclassing from other packages, in which case you'd want to make your overrideable test methods either public or protected. And of course, interface default methods must be public.
> 
> Long story, short: we (the JUnit 5 team) believe in the principle "Less is more", meaning the less you have to type to achieve your goal, the better!

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 5 above)
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: Done! I can't remember in which commit though...

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+    }
+
+    @Test
+    void updateLayout_invalidRequestBodyDefinitionIsBlankAndMetadataIsNull_returns400AndLayoutDoesNotChange()
+        throws Exception {
+        Layout layout = createDefaultLayoutInDatabase();
+
+        LayoutRequestDTO request = new LayoutRequestDTO();
+        request.setDefinition("");
+        request.setMetadata(null);
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [0ccaee4](https://github.com/ScottLogic/finos-vuu/pull/68/commits/0ccaee48ab7fa5ad97a978f7a1aa3659d2051f26)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+                    + "exception is com.fasterxml.jackson.core.JsonParseException: Unrecognized "
+                    + "token 'invalidLayout': was expecting (JSON String, Number, Array, Object "
+                    + "or token 'null', 'true' or 'false')\n"
+                    + " at [Source: (org.springframework.util.StreamUtils$NonClosingInputStream);"
+                    + " line: 1, column: 14]"));
+    }
+
+    @Test
+    void updateLayout_validIdAndValidRequest_returns204AndLayoutHasChanged() throws Exception {
+        Layout layout = createDefaultLayoutInDatabase();
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [324444e](https://github.com/ScottLogic/finos-vuu/pull/68/commits/324444ef4fb336ebe577ddc16bb4dd74818cbd6a)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+        assertThat(updatedLayout.getDefinition())
+            .isEqualTo(layoutRequest.getDefinition());
+        assertThat(updatedLayout.getMetadata().getBaseMetadata().getName())
+            .isEqualTo(layoutRequest.getMetadata().getBaseMetadata().getName());
+        assertThat(updatedLayout.getMetadata().getBaseMetadata().getGroup())
+            .isEqualTo(layoutRequest.getMetadata().getBaseMetadata().getGroup());
+        assertThat(updatedLayout.getMetadata().getBaseMetadata().getScreenshot())
+            .isEqualTo(layoutRequest.getMetadata().getBaseMetadata().getScreenshot());
+        assertThat(updatedLayout.getMetadata().getBaseMetadata().getUser())
+            .isEqualTo(layoutRequest.getMetadata().getBaseMetadata().getUser());
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:267]: [324444e](https://github.com/ScottLogic/finos-vuu/pull/68/commits/324444ef4fb336ebe577ddc16bb4dd74818cbd6a)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+
+    @Test
+    void getMetadata_metadataDoesNotExist_returnsEmptyList() throws Exception {
+        mockMvc.perform(get("/layouts/metadata"))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$").isEmpty());
+    }
+
+    @Test
+    void createLayout_validLayout_returnsCreatedLayoutAndLayoutIsPersisted()
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [5c23160](https://github.com/ScottLogic/finos-vuu/pull/68/commits/5c23160fbbff8b7f94125de8e799518981b9e0d3)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+        mockMvc.perform(get("/layouts/{id}", layoutID))
+            .andExpect(status().isBadRequest())
+            .andExpect(content().string(
+                "Failed to convert value of type 'java.lang.String' to required type 'java.util"
+                    + ".UUID'; nested exception is java.lang.IllegalArgumentException: Invalid "
+                    + "UUID string: invalidUUID"));
+    }
+
+    @Test
+    void getMetadata_metadataExists_returnsMetadata() throws Exception {
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [92ebfab](https://github.com/ScottLogic/finos-vuu/pull/68/commits/92ebfabbc9fee12fae503cdf8b5808028ad4f63e)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+        metadata.setBaseMetadata(baseMetadata);
+
+        layout.setDefinition(defaultDefinition);
+        layout.setMetadata(metadata);
+
+        metadataRepository.save(metadata);
+        Layout createdLayout = layoutRepository.save(layout);
+
+        assertThat(layoutRepository.findById(createdLayout.getId()).orElseThrow())
+            .isEqualTo(layout);
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: What do you think to [5493de8](https://github.com/ScottLogic/finos-vuu/pull/68/commits/5493de82563369f5e88935123c2e78d91d284373)?

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+@AutoConfigureMockMvc
+@Transactional
+@ActiveProfiles("test")
+public class LayoutIntegrationTest {
+
+    private static String defaultDefinition;
+    private static String defaultName;
+    private static String defaultGroup;
+    private static String defaultScreenshot;
+    private static String defaultUser;
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: [474c23e](https://github.com/ScottLogic/finos-vuu/pull/68/commits/474c23e4820e5fb6ce30703df3446f5ba4726161)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/service/LayoutServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    void getLayout_returnsLayout() {
+        when(layoutRepository.findById(layoutId)).thenReturn(Optional.of(layout));
+
+        assertThat(layoutService.getLayout(layoutId)).isEqualTo(layout);
+    }
+
+    @Test
+    void getMetadata_returnsMetadata() {
+        when(metadataService.getMetadata()).thenReturn(List.of(metadata));
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/service/LayoutServiceTest.java]: [3009319](https://github.com/ScottLogic/finos-vuu/pull/68/commits/3009319ce36c9443bca2966cbab6fc3dc7a32a97)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/service/LayoutServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        baseMetadata.setGroup("Test Group");
+        baseMetadata.setScreenshot("Test Screenshot");
+        baseMetadata.setUser("Test User");
+
+        metadata.setId(metadataId);
+        metadata.setBaseMetadata(baseMetadata);
+
+        layout.setId(layoutId);
+        layout.setDefinition("");
+        layout.setMetadata(metadata);
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/service/LayoutServiceTest.java]: [ad5ddeb](https://github.com/ScottLogic/finos-vuu/pull/68/commits/ad5ddeb682c1bc6307916f1be7b52ef26e734187)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+        when(layoutService.createLayout(layoutWithoutIds)).thenReturn(layout.getId());
+        when(layoutService.getLayout(layout.getId())).thenReturn(layout);
+        when(modelMapper.map(layout, LayoutResponseDTO.class)).thenReturn(expectedLayoutResponse);
+
+        assertThat(layoutController.createLayout(layoutRequest))
+            .isEqualTo(expectedLayoutResponse);
+    }
+
+    @Test
+    void updateLayout_callsLayoutService() {
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java]: [bc6e526](https://github.com/ScottLogic/finos-vuu/pull/68/commits/bc6e526d1977b492035df0dcb17445881d108486)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java]: [62d3a90](https://github.com/ScottLogic/finos-vuu/pull/68/commits/62d3a90df4c79f75d807fac430caf3b89962bd7a)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+    @Test
+    void getLayout_layoutExists_returnsLayout() {
+        when(layoutService.getLayout(validLayoutId)).thenReturn(layout);
+        when(modelMapper.map(layout, LayoutResponseDTO.class)).thenReturn(
+            expectedLayoutResponse);
+        assertThat(layoutController.getLayout(validLayoutId)).isEqualTo(expectedLayoutResponse);
+    }
+
+    @Test
+    void getLayout_layoutDoesNotExist_throwsNotFoundAndReturns404() {
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java]: [62d3a90](https://github.com/ScottLogic/finos-vuu/pull/68/commits/62d3a90df4c79f75d807fac430caf3b89962bd7a)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+    private LayoutRequestDTO layoutRequest;
+    private LayoutResponseDTO expectedLayoutResponse;
+    private List<MetadataResponseDTO> expectedMetadataResponse;
+
+    @BeforeEach
+    public void setup() {
+        validLayoutId = UUID.randomUUID();
+        doesNotExistLayoutId = UUID.randomUUID();
+        UUID metadataId = UUID.randomUUID();
+        String layoutDefinition = "Test Definition";
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/controller/LayoutControllerTest.java]: [474c23e](https://github.com/ScottLogic/finos-vuu/pull/68/commits/474c23e4820e5fb6ce30703df3446f5ba4726161)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/service/LayoutService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/service/LayoutService.java]: [fc3cd39](https://github.com/ScottLogic/finos-vuu/pull/68/commits/fc3cd391e14234de0cc3b6f1c49adbe5d488fec7)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/model/Metadata.java [hunk 24]
```diff
@@ -1,16 +1,18 @@
...
 import javax.persistence.GeneratedValue;
 import javax.persistence.GenerationType;
 import javax.persistence.Id;
+
+import lombok.Data;
+
 import lombok.AllArgsConstructor;
 import lombok.Builder;
-import lombok.Data;
 import lombok.NoArgsConstructor;
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/model/Metadata.java]: Done!

**Comment by cfisher-scottlogic** [comment]
> There are quite a few alterations for an "add tests" PR - are these all things that have come up as a result of writing the tests?

I think so, what are you referring to in particular? The exception handling was all implemented in this PR because it was through writing the tests I uncovered what exceptions needed to be handled. That also meant some things needed to change to throw relevant exceptions to handle unhappy-path scenarios. 

Are there any alterations which aren't relevant to this?

> Could also do with a few more "unhappy" paths for the service.

I'm a bit test brain-dead. Can you suggest any unhappy paths I'm missing? I've tried to cover as many unhappy paths as possible with the integration tests, and possibly overlooked unit tests as they felt less valuable.

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: Gone! [f2ca868](https://github.com/ScottLogic/finos-vuu/pull/68/commits/f2ca86873608dc39362b3aa6748e8a9e9de4fb85)

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: I think half of those `[0]`s should be `[1]`s. I'd rather see two distinct entries, but that's not so important.

**Comment by pling-scottlogic** [comment]
> > There are quite a few alterations for an "add tests" PR - are these all things that have come up as a result of writing the tests?
> 
> I think so, what are you referring to in particular? The exception handling was all implemented in this PR because it was through writing the tests I uncovered what exceptions needed to be handled. That also meant some things needed to change to throw relevant exceptions to handle unhappy-path scenarios.
> 
> Are there any alterations which aren't relevant to this?

I didn't see any changes that I wasn't happy with, so I think we're good. Probably mostly exception handling and its fallout.
 
> > Could also do with a few more "unhappy" paths for the service.
> 
> I'm a bit test brain-dead. Can you suggest any unhappy paths I'm missing? I've tried to cover as many unhappy paths as possible with the integration tests, and possibly overlooked unit tests as they felt less cost effective.

I think it's just the case when the delete throws an exception.

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: Oops, my bad. Sorted!
